(feat) O3-3958: Add a select all option for concept answers

NethmiRodrigo opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary
<!-- Please describe what problems your PR addresses. -->
This PR adds the option `Select all` for when users are selecting answers from a concept.

## Screenshots
<!-- Required if you are making UI changes. -->

https://github.com/user-attachments/assets/38ad795a-6d13-4fac-ace5-a48e291e3bd0


## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
https://issues.openmrs.org/browse/O3-3958

## Other
<!-- Anything not covered above -->

ibacher commented:
Code here is fine, but UX question: do you think it maybe makes sense for all possible Concept Answers to be selected by default so that people are basically de-selecting answers they don't want?

NethmiRodrigo commented:
> Code here is good, but UX question: do you think it maybe makes sense for all possible Concept Answers to be selected by default so that people are basically de-selecting answers they don't want?

Makes sense to me!

This made me also think, the reason for having a "Select All" option was for cases where there would be a lot of answers, would it also be good to have this answer dropdown be searchable?

ibacher commented:
> would it also be good to have this answer dropdown be searchable?

Definitely! Especially because people have a bad habit of doing something like adding all ICD-11 codes as an answer...

NethmiRodrigo commented:
> Definitely! Especially because people have a bad habit of doing something like adding all ICD-11 codes as an answer...

Turns out the `FilterableMultiSelect` component doesn't support having `Select All` out of the box...

denniskigen commented on src/components/interactive-builder/modals/question/question-form/rendering-types/inputs/select/select-answers.component.tsx:
```diff
@@ -19,12 +19,31 @@ const SelectAnswers: React.FC = () => {
+        },
+      }));
+    }
+  }, [concept, formField.questionOptions?.answers, setFormField]);
```
Nice work implementing the select-all-by-default suggestion, and having checked the form engine, writing the concept answers into the schema is exactly right since radio/dropdown/multi-select have no runtime fallback to concept answers. One behavior worth fixing before merge: because this effect re-runs whenever the answers list becomes empty (the deps include `formField.questionOptions?.answers`), clearing the selection instantly re-selects everything. The MultiSelect exposes a clear-selection control, so an author who wants a small subset of a large answer list, like the ICD-11 case discussed above, can never clear and start from zero. Could we initialize only when the concept first arrives (for example, track the last-initialized concept uuid in a ref) instead of whenever answers are empty? Guarding against saving with zero answers feels like a separate validation concern rather than something the initializer should enforce.

denniskigen replied:
Addressed in 1139c67. The initializer now runs once per concept, tracked via a ref, so a cleared selection stays cleared. Added a regression test that fails against the previous behavior.

denniskigen commented on src/components/interactive-builder/modals/question/question-form/rendering-types/inputs/select/select-answers.component.tsx:
```diff
@@ -170,7 +202,7 @@ const SelectAnswers: React.FC = () => {
   return (
     <Stack gap={5}>
-      {answerItems.length > 0 && (
+      {answerItems.length > 1 && (
```
This guard makes sense for hiding a dropdown that contains only the select-all item. One edge it introduces: the early-return branch above (concept with no answers plus saved custom answers) does not append the select-all item, so a question with exactly one saved custom answer no longer renders the MultiSelect and that answer cannot be deselected on re-edit. Appending the select-all item in that branch too would fix it, and Carbon sorts `isSelectAll` items to the top regardless of where they appear in the array.

denniskigen replied:
Addressed in 1139c67. The custom-answers-only branch now appends the select-all option as well, so the menu still renders for a single saved custom answer. Added a regression test.

denniskigen commented on src/components/interactive-builder/modals/question/question-form/rendering-types/inputs/select/select-answers.component.tsx:
```diff
@@ -131,7 +152,15 @@ const SelectAnswers: React.FC = () => {
+      ...additionalAnswers,
+      {
+        id: 'select-all',
+        text: 'Select all options',
```
Tiny one: `'Select all options'` should go through translation like the rest of this component, e.g. `t('selectAllOptions', 'Select all options')` with a matching `en.json` entry.

denniskigen replied:
Addressed in 1139c67 with a new selectAllOptions translation key.

denniskigen reviewed (commented):
Took a fresh look at this since it has been open for a while. The approach holds up well: I checked the form engine and `questionOptions.answers` is the sole runtime source of options for radio/dropdown/multi-select, with no fallback to concept answers, so initializing the schema with the concept answers is the right mechanism. Three inline comments below, one behavioral and two small. Also, a rebase is needed before this can land: the component was touched by #403 and #1086, and the test file has since migrated to vitest (#1198), so the updated test needs porting.

NethmiRodrigo committed:
Rebase onto main and adapt tests to vitest and Carbon 1.92

Reapplies the select-all feature on top of current main. The test file
is ported to the vitest-era version of the suite, and the selected-state
assertions now check the answer checkboxes directly because Carbon 1.92
no longer renders the title attribute the previous assertions relied on.

denniskigen committed:
Address review feedback on the select-all behavior

- Initialize the concept's answers once per concept so a cleared
  selection stays cleared instead of being instantly re-selected.
- Append the select-all option in the custom-answers-only branch so a
  question with a single saved custom answer still renders the answers
  menu on re-edit.
- Run the select-all label through i18n with a new selectAllOptions key.
- Add regression tests for the first two behaviors; both fail against
  the previous code.
